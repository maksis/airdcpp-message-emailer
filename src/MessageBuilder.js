
const reduceMessageSummary = (reduced, message) => {
	const timeOffset = (-1) * new Date().getTimezoneOffset() * 60 * 1000;
	const timeString = new Date((message.time * 1e3) + timeOffset).toISOString().slice(-13, -5);
	reduced += `[${timeString}] <${message.from.nick}> ${message.text}\n`;
	return reduced;
};

const hasUnread = (sessionInfo) => {
	const { unread } = sessionInfo.message_counts;
	return unread.bot > 0 || unread.user > 0;
};

const constructSummary = async ({ title, sessionNameGetter }, cache, sessionInfoGetter) => {
	let sessionsInfos = (await Promise.all(Object.keys(cache)
		.map(sessionInfoGetter)))
		.filter(hasUnread);

	if (sessionsInfos.length == 0) {
		return '';
	}

	let ret = '';
	ret += `------- ${title} -------`;
	ret += '\n\n';

	return sessionsInfos.reduce((reduced, sessionInfo) => {
		reduced += `-- ${sessionNameGetter(sessionInfo)} --`;
		reduced += '\n\n';
		
		reduced += cache[sessionInfo.id].reduce(reduceMessageSummary, '');
		reduced += '\n\n';

		return reduced;
	}, ret);
}

export default {
	constructSummary,
};