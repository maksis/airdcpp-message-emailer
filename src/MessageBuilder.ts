import { Message, SessionType, MessageCache, SessionInfoGetter, SessionInfo, SessionId } from './types';

const formatTimeStamp = (time: number) => {
  const timeOffset = (-1) * new Date().getTimezoneOffset() * 60 * 1000;
  return new Date((time * 1e3) + timeOffset).toISOString().slice(-13, -5);
}

const reduceMessageSummary = (reduced: string, message: Message) => {
  const timeString = formatTimeStamp(message.time);
  reduced += `[${timeString}] <${message.from.nick}> ${message.text}\n`;
  return reduced;
};

const hasUnread = (sessionInfo: SessionInfo) => {
  const { unread } = sessionInfo.message_counts;
  return unread.bot > 0 || unread.user > 0;
};

const mapSessionInfo = async (sessionId: SessionId, sessionInfoGetter: SessionInfoGetter) => {
  try {
    return await sessionInfoGetter(sessionId);
  } catch (e) {
    // Session removed, assume that is has been read
  }

  return null;
};

const constructSummary = async ({ title, sessionNameGetter }: SessionType, cache: MessageCache, sessionInfoGetter: SessionInfoGetter) => {
  if (!cache) {
    return '';
  }

  const sessionsInfos = (await Promise.all(Object.keys(cache)
    .map(sessionId => mapSessionInfo(sessionId, sessionInfoGetter))))
    .filter(sessionInfo => !!sessionInfo)
    .filter(hasUnread);

  if (sessionsInfos.length == 0) {
    return '';
  }

  let ret = '';
  ret += `------- ${title} -------`;
  ret += '\n\n';

  return sessionsInfos.reduce((reduced: string, sessionInfo: any) => {
    reduced += `-- ${sessionNameGetter(sessionInfo)} --`;
    reduced += '\n\n';
    
    reduced += cache[sessionInfo.id].reduce(reduceMessageSummary, '');
    reduced += '\n\n';

    return reduced;
  }, ret);
}

export default {
  constructSummary,
};
