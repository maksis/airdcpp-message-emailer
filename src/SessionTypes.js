
export default {
	hub: {
		title: 'Hub messages',
		sessionNameGetter: sessionInfo => sessionInfo.identity.name,
	},
	privateChat: {
		title: 'Private messages',
		sessionNameGetter: sessionInfo => sessionInfo.user.nicks,
	}
}