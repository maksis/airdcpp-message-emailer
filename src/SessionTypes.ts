import { HubSession, PrivateChatSession, SessionType } from "./types";

const SessionTypes = {
  hub: {
    title: 'Hub messages',
    sessionNameGetter: (sessionInfo: HubSession) => sessionInfo.identity.name,
  } as SessionType,
  privateChat: {
    title: 'Private messages',
    sessionNameGetter: (sessionInfo: PrivateChatSession) => sessionInfo.user.nicks,
  } as SessionType,
};

export default SessionTypes;
