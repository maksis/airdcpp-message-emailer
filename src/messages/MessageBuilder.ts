import { ChatMessage, SessionParser, MessageCache, SessionInfoFetcher, SessionInfo, SessionId } from '../types';

const formatTimeStamp = (time: number) => {
  const timeOffset = (-1) * new Date().getTimezoneOffset() * 60 * 1000;
  return new Date((time * 1e3) + timeOffset).toISOString().slice(-13, -5);
}

const reduceMessageSummary = (reduced: string, message: ChatMessage) => {
  const timeString = formatTimeStamp(message.time);
  reduced += `[${timeString}] <${message.from.nick}> ${message.text}\n`;
  return reduced;
};

const hasUnread = (sessionInfo: SessionInfo) => {
  const { unread } = sessionInfo.message_counts;
  return unread.bot > 0 || unread.user > 0;
};

const mapSessionInfo = async (sessionId: SessionId, SessionInfoFetcher: SessionInfoFetcher<SessionInfo>) => {
  try {
    // Skip the session cache to get the latest read status
    const session = await SessionInfoFetcher(sessionId, true);
    return session;
  } catch (e) {
    // Session removed, assume that is has been read
  }

  return null;
};

const constructSummary = async ({ title, sessionNameGetter }: SessionParser, cache: MessageCache, SessionInfoFetcher: SessionInfoFetcher<SessionInfo>) => {
  if (!cache) {
    return '';
  }

  const sessionsInfos = (await Promise.all(Object.keys(cache)
    .map(sessionId => mapSessionInfo(sessionId, SessionInfoFetcher))))
    .filter(sessionInfo => !!sessionInfo)
    .filter(hasUnread);

  if (sessionsInfos.length == 0) {
    return '';
  }

  let ret = '';
  ret += `------- ${title} -------`;
  ret += '\n\n';

  return sessionsInfos.reduce((reduced: string, sessionInfo: any) => {
    reduced += `-- ${sessionNameGetter(sessionInfo)} --`;
    reduced += '\n\n';
    
    reduced += cache[sessionInfo.id].reduce(reduceMessageSummary, '');
    reduced += '\n\n';

    return reduced;
  }, ret);
}

export default {
  constructSummary,
};
