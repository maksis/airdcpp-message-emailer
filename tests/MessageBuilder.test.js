import MessageBuilder from '../src/MessageBuilder';
import SessionTypes from '../src/SessionTypes';


describe('MessageBuilder', () => {
  const getTimezoneOffset = Date.prototype.getTimezoneOffset;

  beforeEach(() => {
    Date.prototype.getTimezoneOffset = () => {
      return -180;
    }
  });

  afterEach(() => {
    Date.prototype.getTimezoneOffset = getTimezoneOffset;
  });

  const message1 = {
    from: {
      nick: 'TestNick1'
    },
    text: 'TestMessage1',
    time: 1494434082,
  };

  const message2 = {
    from: {
      nick: 'TestNick2'
    },
    text: 'TestMessage2',
    time: 1494435082,
  };

  const privateChatId1 = 'VZILDK44YRK37UJBED2KLZSORDWRKMLDIYIRULY';
  const privateChatId2 = 'AZILDK44YRK37UJFED2KLZSORDWRKMLDIYIRULY';

  const sessionInfo1 = {
    id: privateChatId1,
    user: {
      nicks: '[100]User1',
    },
    message_counts: {
      unread: {
        bot: 0,
        status: 0,
        user: 2,
      },
    },
  };

  const sessionInfo2 = {
    id: privateChatId2,
    user: {
      nicks: '[100]User2',
    },
    message_counts: {
      unread: {
        bot: 0,
        status: 0,
        user: 1,
      },
    },
  };


  const readSessionInfo = {
    user: {
      nicks: '[100]SeenUser',
    },
    message_counts: {
      unread: {
        bot: 0,
        status: 0,
        user: 0,
      },
    },
  };

  const cache = {
    [privateChatId1]: [
      message1,
      message2,
    ], 
    [privateChatId2]: [
      message1,
      message2,
    ]
  };

  test('should format messages', async () => {
    const sessionInfoGetter = sessionId => sessionId === privateChatId1 ? sessionInfo1 : sessionInfo2;

    const summary = await MessageBuilder.constructSummary(SessionTypes.privateChat, cache, sessionInfoGetter);
    expect(summary).toMatchSnapshot();
  });

  test('should skip read sessions', async () => {
    const sessionInfoGetter = (_sessionId) => readSessionInfo

    const summary = await MessageBuilder.constructSummary(SessionTypes.privateChat, cache, sessionInfoGetter);
    expect(summary).toEqual('');
  });

  test('should handle removed sessions', async () => {
    const sessionInfoGetter = (_sessionId) => Promise.reject('Session removed');

    const summary = await MessageBuilder.constructSummary(SessionTypes.privateChat, cache, sessionInfoGetter);
    expect(summary).toEqual('');
  });
});