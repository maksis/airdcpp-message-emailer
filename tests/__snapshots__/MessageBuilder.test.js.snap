// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageBuilder should format messages 1`] = `
"------- Private messages -------

-- [100]User1 --

[19:34:42] <TestNick1> TestMessage1
[19:51:22] <TestNick2> TestMessage2


-- [100]User2 --

[19:34:42] <TestNick1> TestMessage1
[19:51:22] <TestNick2> TestMessage2


"
`;
